### ISSUE TITLE
[MASSEMBLY-683] dependencySet includes different versions of same dependency when there is a pom relocation involved

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-assembly-plugin#889]
**[Claus Schrammel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=f05fk)** opened **[MASSEMBLY-683](https://issues.apache.org/jira/browse/MASSEMBLY-683?redirect=false)** and commented

When dependencySets are resolved, the assembly plugin handles different versions of the same artifact and provides the same result as Maven built-in dependency resolution does.

But it fails if the artifact is relocated in its pom. The attached example shows this for xerces:xerces.

The app depends on lib-a, which itself depends on xerces:xerces:1.4.4.
The app also depends on lib-b, which itself depends on xerces:xerces:2.0.2, but that is relocated to xerces:xercesImpl.

So the dependencySet on app contains xerces-1.4.4 **and** xercesImpl-2.0.2, causing problems on the classpath of the actual application.

The demo contains a standard case with 2 different log4j version which is working fine.

The demo also uses the maven-dependency-plugin that is able to resolve the dependencies correctly. Only xerces-1.4.4 goes into the target/dependency folder.

---

**Affects:** 2.4

**Attachments:**
- [acme.demo.zip](https://issues.apache.org/jira/secure/attachment/12714397/acme.demo.zip) (_27.36 kB_)

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

The point is that for xerces-1.4.4 no relocation enty in central exists whereas for xerces-2.0.2 it does.
This can't be solved by maven-assembly-plugin nor by maven.

In that case the groupId/artifactId's are different in the poms which means having both as a result. Such a warning about relocated artifacts should be fixed.

Based on the above I can't see a wrong behaviour. So i will closes the issue. If you have any objections please don't hesitate to reopen the issue.

**Comment by jira-importer** [comment]
**[Claus Schrammel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=f05fk)** commented

I am sorry that I was not clear enough in my original bug report. The first line may be misleading. **Normally** they provide the same result.

But actually the assembly plugin and Maven (maven-dependency-plugin) provide **different** results for the given xerces issue.

Please try the attached example and compare the output of the assembly plugin (lib folder within target/app-0.0.1-SNAPSHOT-package.zip) with the output of the maven-dependency-plugin (target/dependency).

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Hi, now i got your point sorry was a misunderstanding on my site.

Lets dive into it. If i use maven-dependency-plugin:2.4 you will see it like this:

```
$ mvn org.apache.maven.plugins:maven-dependency-plugin:2.4:tree 
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building app 0.0.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-dependency-plugin:2.4:tree (default-cli) @ app ---
[INFO] acme.demo:app:jar:0.0.1-SNAPSHOT
[INFO] +- junit:junit:jar:4.11:test
[INFO] |  \- org.hamcrest:hamcrest-core:jar:1.3:test
[INFO] +- acme.demo:lib-a:jar:0.0.1-SNAPSHOT:compile
[INFO] |  +- log4j:log4j:jar:1.2.16:compile
[INFO] |  \- xerces:xerces:jar:1.4.4:compile
[INFO] \- acme.demo:lib-b:jar:0.0.1-SNAPSHOT:compile
[INFO]    \- xerces:xercesImpl:jar:2.0.2:compile
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
```

But if you use maven-dependency-plugin:2.5 you will get the following:

```
$ mvn org.apache.maven.plugins:maven-dependency-plugin:2.5:tree 
[INFO] Scanning for projects...
[INFO]                                                                         
...
[INFO] --- maven-dependency-plugin:2.5:tree (default-cli) @ app ---
[INFO] acme.demo:app:jar:0.0.1-SNAPSHOT
[INFO] +- junit:junit:jar:4.11:test
[INFO] |  \- org.hamcrest:hamcrest-core:jar:1.3:test
[INFO] +- acme.demo:lib-a:jar:0.0.1-SNAPSHOT:compile
[INFO] |  +- log4j:log4j:jar:1.2.16:compile
[INFO] |  \- xerces:xerces:jar:1.4.4:compile
[INFO] \- acme.demo:lib-b:jar:0.0.1-SNAPSHOT:compile
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
```

If you use maven-dependency-plugin:2.6?2.8 you will get the same results which is related to [MSHARED-167](http://jira.codehaus.org/browse/MSHARED-167)

**Comment by jira-importer** [comment]
**[Claus Schrammel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=f05fk)** commented

Exactly. It seems, that has been fixed in maven-dependency-plugin:2.5.

Interestingly, when calling copy-dependencies, none of the maven-dependency-plugin versions copies xercesImpl-2.0.2.

Not even maven-dependency-plugin:2.4, although it shows it in the dependency tree, it does not copy it:

```
C:\tmp\acme.demo\acme.demo.app>mvn org.apache.maven.plugins:maven-dependency-plugin:2.4:copy-dependencies
[INFO] Scanning for projects...
[INFO]
...
[INFO] --- maven-dependency-plugin:2.4:copy-dependencies (default-cli) @ app ---
[INFO] Copying lib-a-0.0.1-SNAPSHOT.jar to C:\tmp\acme.demo\acme.demo.app\target\dependency\lib-a-0.0.1-SNAPSHOT.jar
[INFO] Copying hamcrest-core-1.3.jar to C:\tmp\acme.demo\acme.demo.app\target\dependency\hamcrest-core-1.3.jar
[INFO] Copying junit-4.11.jar to C:\tmp\acme.demo\acme.demo.app\target\dependency\junit-4.11.jar
[INFO] Copying xerces-1.4.4.jar to C:\tmp\acme.demo\acme.demo.app\target\dependency\xerces-1.4.4.jar
[INFO] Copying log4j-1.2.16.jar to C:\tmp\acme.demo\acme.demo.app\target\dependency\log4j-1.2.16.jar
[INFO] Copying lib-b-0.0.1-SNAPSHOT.jar to C:\tmp\acme.demo\acme.demo.app\target\dependency\lib-b-0.0.1-SNAPSHOT.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

But maven-assembly-plugin packs xercesImpl-2.0.2 into the zip file:

```
C:\tmp\acme.demo\acme.demo.app>mvn package
[INFO] Scanning for projects...
[INFO]
...
[INFO] --- maven-dependency-plugin:2.7:copy-dependencies (copy-dependency-jars) @ app ---
[INFO] Copying lib-a-0.0.1-SNAPSHOT.jar to C:\tmp\acme.demo\acme.demo.app\target\dependency\lib-a-0.0.1-SNAPSHOT.jar
[INFO] Copying xerces-1.4.4.jar to C:\tmp\acme.demo\acme.demo.app\target\dependency\xerces-1.4.4.jar
[INFO] Copying log4j-1.2.16.jar to C:\tmp\acme.demo\acme.demo.app\target\dependency\log4j-1.2.16.jar
[INFO] Copying lib-b-0.0.1-SNAPSHOT.jar to C:\tmp\acme.demo\acme.demo.app\target\dependency\lib-b-0.0.1-SNAPSHOT.jar
[INFO]
[INFO] --- maven-assembly-plugin:2.4:single (package-zip) @ app ---
[INFO] Reading assembly descriptor: package-zip.xml
[INFO] Building zip: C:\tmp\acme.demo\acme.demo.app\target\app-0.0.1-SNAPSHOT-package.zip
[DEBUG] adding directory app-0.0.1-SNAPSHOT/
[DEBUG] adding entry app-0.0.1-SNAPSHOT/README.txt
[DEBUG] adding directory app-0.0.1-SNAPSHOT/lib/
[DEBUG] adding entry app-0.0.1-SNAPSHOT/lib/lib-a-0.0.1-SNAPSHOT.jar
[DEBUG] adding entry app-0.0.1-SNAPSHOT/lib/log4j-1.2.16.jar
[DEBUG] adding entry app-0.0.1-SNAPSHOT/lib/xerces-1.4.4.jar
[DEBUG] adding entry app-0.0.1-SNAPSHOT/lib/lib-b-0.0.1-SNAPSHOT.jar
[DEBUG] adding entry app-0.0.1-SNAPSHOT/lib/xercesImpl-2.0.2.jar
[DEBUG] adding entry app-0.0.1-SNAPSHOT/lib/app-0.0.1-SNAPSHOT.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This issue has been auto closed because it has been inactive for a long period of time. If you think this issue still applies, retest your problem with the most recent version of Maven and the affected component, reopen and post your results.
